File: solutions/julia/matching-brackets/1/matching-brackets.jl

function is_left(ch)
    ch == '(' || ch == '{' || ch=='['
end

function is_right(ch)
    ch == ')' || ch == '}' || ch==']'
end 

function matches(l, r)
   l == '(' && r == ')' || l == '{' && r == '}' || l == '[' && r == ']'
end

function matching_brackets( input )
    lifo = Char[] 
    for ch in input
        if is_left(ch)
            push!(lifo, ch)
        elseif is_right(ch)
            if 0==length(lifo)
                return false
            elseif matches(pop!(lifo),ch)
                continue
            else
                return false
            end        
        end
    end
    return 0==length(lifo)
end

File: solutions/kotlin/matching-brackets/1/src/main/kotlin/MatchingBrackets.kt
object MatchingBrackets {

    private fun isLeftDelim(ch : Char): Boolean =
        ch=='(' || ch=='[' || ch=='{'

    private fun isRightDelim(ch : Char): Boolean =
        ch==')' || ch==']' || ch=='}'

    private fun closes(l: Char, r : Char) : Boolean = 
        (l == '(' && r == ')') ||
        (l == '[' && r == ']') ||
        (l == '{' && r == '}')     

    
    fun isValid(input: String): Boolean {
        var lifo = ArrayList<Char>()
        input.forEach { ch ->
            if (isLeftDelim(ch)) {
                lifo.add(ch)                
            } 
            if (isRightDelim(ch)) {
                if (lifo.isEmpty())
                    return false
                if (closes(lifo.last(),ch)) 
                    lifo.removeLast()
                else return false
            }
        }
        return lifo.isEmpty()
    }
}
